src/vsr/message_header.zig

```diff
@@ -0,0 +1,1430 @@
+    /// detecting whether a session has been evicted is solved by the session number.
+    client: u128,
+    /// When operation=register, this is zero.
+    /// When operation≠register, this is the commit number of register.
+    session: u64 = 0,
+    /// Only nonzero during AOF recovery.
+    /// TODO: Use this for bulk-import to state machine?
+    timestamp: u64 = 0,
+    /// Each request is given a number by the client and later requests must have larger numbers
+    /// than earlier ones. The request number is used by the replicas to avoid running requests
+    /// more than once; it is also used by the client to discard duplicate replies to its requests.
+    /// A client is allowed to have at most one request inflight at a time.
+    request: u32,
+    operation: Operation,
+    reserved: [123]u8 = [_]u8{0} ** 123,
+
+    pub fn frame(header: *@This()) *Header {
+        return std.mem.bytesAsValue(Header, std.mem.asBytes(header));
+    }
+
+    pub fn frame_const(header: *const @This()) *const Header {
+        return std.mem.bytesAsValue(Header, std.mem.asBytes(header));
+    }
+
+    fn invalid(self: *const @This()) ?[]const u8 {
+        assert(self.command == .request);
+        if (self.replica != 0) return "replica != 0";
+        if (self.client == 0) return "client == 0";
+        if (self.timestamp != 0 and !constants.aof_recovery) return "timestamp != 0";
+        switch (self.operation) {
+            .reserved => return "operation == .reserved",
+            .root => return "operation == .root",
+            .register => {
+                // The first request a client makes must be to register with the cluster:
+                if (self.parent != 0) return "register: parent != 0";
+                if (self.session != 0) return "register: session != 0";
+                if (self.request != 0) return "register: request != 0";
+                // The .register operation carries no payload:
+                if (self.checksum_body != checksum_body_empty) {
+                    return "register: checksum_body != expected";
+                }
+                if (self.size != @sizeOf(Header)) return "size != @sizeOf(Header)";
+            },
+            else => {
+                if (self.operation == .reconfigure) {
+                    if (self.size != @sizeOf(Header) + @sizeOf(vsr.ReconfigurationRequest)) {
+                        return "size != @sizeOf(Header) + @sizeOf(ReconfigurationRequest)";
+                    }
+                } else if (@intFromEnum(self.operation) < constants.vsr_operations_reserved) {
+                    return "operation is reserved";
+                }
+                // Thereafter, the client must provide the session number:
+                // These requests should set `parent` to the `checksum` of the previous reply.
+                if (self.session == 0) return "session == 0";
+                if (self.request == 0) return "request == 0";
+            },
```

Comment:
Let's add comment here that we check that the operation is a valid enum variant in replia, as that requires knowing the StateMachine type.